Load custom QT themes from yuzu directory

flodavid opened the pull request:
- User should put their themes in the "qt_themes" subfolder of Yuzu
- Each theme goes in its sub-directory
  - The name should be a lowercase string separated by underscores (snake case)
- If directory name contains "dark", the theme will be considered dark
- The directory must contain a style.qss file

flodavid commented:
Example of custom theme in "pink_dark" folder
![Screenshot from 2024-01-21 21-18-19](https://github.com/yuzu-emu/yuzu/assets/11854849/2c6c9555-b36b-4f24-b9c5-37f4fde7cf0d)

flodavid commented:
Windows compilation should be fixed, but theme switching is not consistent for colorful themes

liamwhite commented:
Can you fix the format job so we can get a full CI run

FearlessTobi commented on src/yuzu/configuration/configure_ui.cpp:
```diff
@@ -106,11 +106,28 @@ ConfigureUi::ConfigureUi(Core::System& system_, QWidget* parent)
     }
 
+    // Add custom styles stored in yuzu directory
+    QDir local_dir(QString::fromStdString(
```
```suggestion
    const QDir local_dir(QString::fromStdString(
```

FearlessTobi commented on src/yuzu/configuration/configure_ui.cpp:
```diff
@@ -106,11 +106,28 @@ ConfigureUi::ConfigureUi(Core::System& system_, QWidget* parent)
+        // Split at _ and capitalize words in name
+        QStringList cased;
+        for (QString word : theme_name.split(QChar::fromLatin1('_'))) {
+            cased << word.at(0).toUpper() + word.mid(1);
```
```suggestion
            cased.append(word.at(0).toUpper() + word.mid(1));
```

FearlessTobi commented on src/yuzu/configuration/configure_ui.cpp:
```diff
@@ -106,11 +106,28 @@ ConfigureUi::ConfigureUi(Core::System& system_, QWidget* parent)
+        }
+        theme_name = cased.join(QChar::fromLatin1(' '));
+
+        theme_name += QString::fromStdString(" (") + tr("Custom") + QString::fromStdString(")");
```
This should be a QStringLiteral with the translated string being an `.arg`

FearlessTobi commented on src/yuzu/configuration/configure_ui.cpp:
```diff
@@ -106,11 +106,28 @@ ConfigureUi::ConfigureUi(Core::System& system_, QWidget* parent)
+    // Add custom styles stored in yuzu directory
+    QDir local_dir(QString::fromStdString(
+        Common::FS::GetYuzuPathString(Common::FS::YuzuPath::ThemesDir)));
+    for (QString& theme_dir : local_dir.entryList(QDir::NoDot | QDir::NoDotDot | QDir::Dirs)) {
```
```suggestion
    for (const QString& theme_dir : local_dir.entryList(QDir::NoDot | QDir::NoDotDot | QDir::Dirs)) {
```

FearlessTobi commented on src/yuzu/main.cpp:
```diff
@@ -5038,26 +5037,53 @@ void GMainWindow::UpdateUITheme() {
-            LOG_ERROR(Frontend, "Unable to open style \"{}\", fallback to the default theme",
+        if (f.open(QFile::ReadOnly | QFile::Text)) {
+            LOG_INFO(Frontend, "Loading style from app: {}", theme_uri.toStdString());
+            LoadStylesheet(&f);
```
Please return early here, so we have less nesting of if-statements

FearlessTobi commented on src/yuzu/main.cpp:
```diff
@@ -5038,26 +5037,53 @@ void GMainWindow::UpdateUITheme() {
+                current_theme.toStdString() / "style.qss";
+            QString theme_stringpath = QString::fromStdString(theme_path.string());
+
+            QFile local_f(theme_stringpath);
```
QFile takes std::filesystem::paths according to the documentation. Also please improve the variable name

flodavid replied:
It accepts std::filesystem::paths only with Qt 6 sadly

FearlessTobi commented on src/yuzu/main.cpp:
```diff
@@ -5038,26 +5037,53 @@ void GMainWindow::UpdateUITheme() {
+            QFile local_f(theme_stringpath);
+            if (local_f.open(QFile::ReadOnly | QFile::Text)) {
+                LOG_INFO(Frontend, "Loading style in: {}", theme_path.string());
+                LoadStylesheet(&local_f);
```
Please return early here, so we have less nesting of if-statements

FearlessTobi commented on src/yuzu/main.cpp:
```diff
@@ -5038,26 +5037,53 @@ void GMainWindow::UpdateUITheme() {
+
+                QFile default_file(theme_uri);
+                if (f.open(QFile::ReadOnly | QFile::Text)) {
+                    LoadStylesheet(&default_file);
```
Please return early here, so we have less nesting of if-statements

FearlessTobi commented on src/yuzu/main.cpp:
```diff
@@ -5038,26 +5037,53 @@ void GMainWindow::UpdateUITheme() {
-        qApp->setStyleSheet({});
-        setStyleSheet({});
-    }
+void GMainWindow::LoadStylesheet(QFile* f) {
```
```suggestion
void GMainWindow::LoadStylesheet(const QFile& file) {
```

flodavid commented:
Thank you for your feedback, its appreciated. All issues should be fixed now. I reworked the loading of stylesheet to use a string instead of QFile. Its simplifies things and its clearer why we are returning early.
I added the feature allowing to override icons in addition to the stylesheet. I also added "midnight" as a word that defines the theme as dark.

I am still working on this PR to see if I can improve the handling of automatic dark/light mode switching.

liamwhite commented:
Please fix windows build

flodavid commented:
I removed logs or reduced level. Add details to commit messages.

This PR has been tested on Pop!_OS 22.04 and Windows 11.
New dark mode option : 
![Screenshot from 2024-02-14 19-28-39](https://github.com/yuzu-emu/yuzu/assets/11854849/70278308-334b-4ac1-abe8-0e036fa584ce)

flodavid commented:
I fixed the last commit, that broke Auto dark mode on Linux.
I also added a new commit that fixes the style of the qdarkstyle and qdarkstyle_midnight_blue sliders, they were cropped for the ZR and ZL threshold, and the volume level: 
| Before| After |
|-----------|---------|
| ![Screenshot from 2024-02-23 16-34-21](https://github.com/yuzu-emu/yuzu/assets/11854849/33177046-f0c1-42a6-ac0d-6391ef789664) | ![image](https://github.com/yuzu-emu/yuzu/assets/11854849/3dec0294-98ea-4bfc-bd2a-0049495a7381) |

Kevnkkm commented:
Does this consider the changes made in #12852 and #12930?
